src/Compilers/CSharp/Portable/Lowering/LocalRewriter/LocalRewriter_TupleBinaryOperator.cs

```diff
@@ -129,12 +143,20 @@ private BoundExpression ReplaceTerminalElementsWithTemps(BoundExpression expr, T
             BoundExpression leftHasValue;
             BoundExpression leftValue;
 
-            // Note: left and right are either temps or `null`, so we don't have detailed information to tell us a nullable always has a value
             var isLeftNullable = left.Kind != BoundKind.TupleLiteral && left.Type.IsNullableType();
             if (isLeftNullable)
             {
-                leftHasValue = MakeHasValueTemp(left, temps, outerEffects);
-                leftValue = MakeValueOrDefaultTemp(left, temps, innerEffects);
+                BoundExpression value = NullableAlwaysHasValue(left);
+                if (value is null)
+                {
+                    leftHasValue = MakeHasValueTemp(left, temps, outerEffects);
+                    leftValue = MakeValueOrDefaultTemp(left, temps, innerEffects);
+                }
+                else
+                {
+                    leftHasValue = MakeBooleanConstant(left.Syntax, true);
+                    leftValue = value;
```

Comment:
>leftValue = value; [](start = 20, length = 18)

```isLeftNullable = false;``` ?

Reply:
`isLeftNullable` isn't used below.

Reply:
> isLeftNullable isn't used below.

It is used on line 201 below ```if (!isLeftNullable && !isRightNullable)```

---
In reply to: [176383909](https://github.com/dotnet/roslyn/pull/25644#discussion_r176383909) [](ancestors = 176383909)